security/pkg/server/ca/node_auth.go

```diff
@@ -16,27 +16,157 @@ package ca
 
 import (
 	"fmt"
+	"sync"
 
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 
+	"istio.io/istio/pkg/cluster"
 	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
 	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/multicluster"
 	"istio.io/istio/pkg/security"
 	"istio.io/istio/pkg/spiffe"
 )
 
+func isZtunnelPod(uobj any) bool {
+	obj, ok := uobj.(controllers.Object)
+	if !ok {
+		return false
+	}
+	v, ok := obj.GetLabels()["app"]
+	if !ok {
+		return false
+	}
+	return v == "ztunnel"
+}
+
 // NodeAuthorizer is a component that implements a subset of Kubernetes Node Authorization
 // (https://kubernetes.io/docs/reference/access-authn-authz/node/) for Istio CA. Specifically, it
 // validates that a node proxy which requests certificates for workloads on its own node is requesting
 // valid identities which run on that node (rather than arbitrary ones).
-type NodeAuthorizer struct {
+type NodeAuthorizer interface {
+	authenticateImpersonation(caller security.KubernetesInfo, requestedIdentityString string) error
+}
+
+// MulticlusterNodeAuthorizor is an implementation of NodeAuthorizer for multi-cluster environmnets.
+// It should be used when ENABLE_MULTICLUSTER_NODE_AUTHORIZER is set, which means node authorizations
+// from remote clusters could also be processed (https://docs.google.com/document/d/10uf4EvUVif4xGeCYQydaKh9Yaz9wpysao7gyLewJY2Q)
+// It is responsible for maintaining an index of node authenticators, one per cluster.
+// node authorizations from one cluster will be forwarded to the node authorizer for the same cluster.
+type MulticlusterNodeAuthorizor struct {
+	remoteNodeAuthenticators map[cluster.ID]*ClusterNodeAuthorizer
+	ztunnelPodsClient        map[cluster.ID]kclient.Client[*v1.Pod]
+	ztunnelIndex             sync.Map
+	m                        sync.Mutex // protects remoteKubeControllers
+	filter                   func(t any) bool
+	trustedNodeAccounts      map[types.NamespacedName]struct{}
+}
+
+func NewMulticlusterNodeAuthenticator(filter func(t any) bool, trustedNodeAccounts map[types.NamespacedName]struct{},
```

Comment:
use func alias instead

Reply:
Thanks, now use the namespace.DiscoveryFilter func alias.